File: backend/src/main/kotlin/com/alpha/domain/repository/RegionRepository.kt
package com.alpha.domain.repository

import com.alpha.domain.entity.RegionEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.util.*

interface RegionRepository : JpaRepository<RegionEntity, UUID> {
    fun findByCode(code: String): RegionEntity?
    fun findByCountry(country: String): List<RegionEntity>
    fun findByIsActiveTrue(): List<RegionEntity>
    fun existsByCode(code: String): Boolean
    
    @Query("SELECT r FROM RegionEntity r WHERE LOWER(r.name) LIKE LOWER(CONCAT('%', :search, '%')) OR LOWER(r.code) LIKE LOWER(CONCAT('%', :search, '%')) OR LOWER(r.country) LIKE LOWER(CONCAT('%', :search, '%'))")
    fun search(@Param("search") search: String): List<RegionEntity>
}

File: backend-scala-zio/src/main/scala/com/alpha/repository/RegionRepository.scala
package com.alpha.repository

import zio.*
import com.alpha.domain.model.*
import java.util.UUID
import io.getquill.*

trait RegionRepository:
  def findAll: Task[List[Region]]
  def findById(id: UUID): Task[Option[Region]]
  def findByCode(code: String): Task[Option[Region]]
  def findByCountry(country: String): Task[List[Region]]
  def findByIsActiveTrue: Task[List[Region]]
  def existsByCode(code: String): Task[Boolean]
  def search(query: String): Task[List[Region]]
  def create(region: Region): Task[UUID]
  def update(region: Region): Task[Int]
  def delete(id: UUID): Task[Int]

object RegionRepository:
  val layer: ZLayer[PostgresCtx, Nothing, RegionRepository] = 
    ZLayer.fromFunction(new RegionRepositoryImpl(_))

class RegionRepositoryImpl(ctx: PostgresCtx) extends RegionRepository:

  import ctx.*

  override def findAll: Task[List[Region]] = 
    ZIO.attempt:
      run(query[Region]).toList

  override def findById(id: UUID): Task[Option[Region]] = 
    ZIO.attempt:
      run(query[Region].filter(_.id == lift(id))).headOption

  override def findByCode(code: String): Task[Option[Region]] = 
    ZIO.attempt:
      run(query[Region].filter(_.code == lift(code))).headOption

  override def findByCountry(country: String): Task[List[Region]] = 
    ZIO.attempt:
      run(query[Region].filter(_.country == lift(country))).toList

  override def findByIsActiveTrue: Task[List[Region]] = 
    ZIO.attempt:
      run(query[Region].filter(_.isActive == lift(true))).toList

  override def existsByCode(code: String): Task[Boolean] = 
    ZIO.attempt:
      run(query[Region].filter(_.code == lift(code)).nonEmpty)

  override def search(queryStr: String): Task[List[Region]] = 
    ZIO.attempt:
      val pattern = "%" + queryStr + "%"
      run(query[Region]
        .filter(r => r.name.like(lift(pattern)) || r.code.like(lift(pattern)))
      ).toList

  override def create(region: Region): Task[UUID] = 
    ZIO.attempt:
      run(query[Region].insertValue(lift(region)))
      region.id

  override def update(region: Region): Task[Int] = 
    ZIO.attempt:
      run(query[Region]
        .filter(_.id == lift(region.id))
        .updateValue(lift(region)))
      1

  override def delete(id: UUID): Task[Int] = 
    ZIO.attempt:
      run(query[Region].filter(_.id == lift(id)).delete)
      1
